### DISCUSSION TITLE
VS Code: Custom Components support (custom components.json)

### DISCUSSION THREAD

**Discussion by mikoskinen** [apache/camel-karavan#1191] [Ideas]
Would it be possible to have a support for using a custom components.json in addition to the one provided by the VS Code extension (C:\Users\xxxvscode\extensions\camel-karavan.karavan-4.4.0\components\components.json)?

The custom file could be provided through the extension's setting. There's the Karavan.supportedComponents -setting, but afaik this works like a filter. My suggestion is more like a Karavan.additionalComponents which would point to some json-file. This json file should be in the same format as the components.json.

**Use case**

We would like add some custom components into the GUI. We're happy if the designer understands these components as we run the actual integrations separately (not through the GUI).

We tested that manually editing the built-in components.json provided the end result we liked. But instead of editing the files provided by an extension, it would be nice if we could provide the file through settings.

**Optional implementation**

Allow the developer to override the default components.json -file by providing a path to the components json file. So instead of the extension having to "merge" two files, it would still only read one file but this time it would be read from a path, if one is set. This is a simpler solution, I think. 

What do you think?

**Comment by davsclaus** [comment]
Yes I think that if we can make adding custom components easier to be seen as first-class into Karavan (both VSCode and web) would be great.

Also if this can be documented, eg if you can help create a .adoc file describing how to do this.

**Comment by mgubaidullin** [comment]
I would suggest that custom Component is a bit more than just json.

Can camel-jbang define and download jar and it's dependencies? Where get this info? And they are in private maven (otherwise component might be just contributed).

I would propose to use custom Kamelet instead. It supports all the above
